Add Custom Date Range to Gantt Charts Part 2

paldeep-singh opened the pull request:
## :bookmark_tabs: Summary

This PR is a duplicate of #4563, to resolve  [#1530](https://github.com/mermaid-js/mermaid/issues/1530). I am slightly desperate for this feature and did not want to trouble @AlexMooney for permission to fork his repo so I have simply copied his changes over into a new branch. Big thanks to him for letting me take over and for doing most of the work! 

Most of the changes included in this PR are Alex's work, I have only done the following:
- Changed logic for determining start date to take an end-only date range into account
  - This fixes the failures on the e2e tests for `Actual date`, `Only start` and `Only end` that Alex wrote in response to the [one](https://github.com/mermaid-js/mermaid/pull/4563/files#r1249319589) outstanding comment on his PR.
  - I have also added the documentation for verifying snapshots to the contribution docs, and had a look at the snapshots in Cypress to verify them (would appreciate a second set of eyes though since I this is my first contribution).
- Fixed the failing ganttDb unit test

Resolves #1530 

## :straight_ruler: Design Decisions

[x] Remove all tasks which are completely outside of the custom range
[x] Force the x-range on the plot to match the custom range (though there might be some rounding)
[x] Allow setting an absolute range dateRange :2023-06-01, 2023-07-01
[ ] Allow setting a relative range dateRange :today - 7d, today + 30d (Alex suggested leaving this for future PRs and I tend to agree)
[x] Allow setting only lower bound dateRange :2023-06-01
[x] Allow setting only upper bound dateRange :,2023-07-01 👈 kind of strange; proposals welcomed
[x] Tests for the range values
[x] Manual testing with some screen shots (As mentioned above, would appreciate a second set of eyes since I this is my first contribution)
[x] Remove backup logic for parsing date with `new Date()` if dayjs determines date provided is invalid.
[x] Set default for `dateFormat` to `YYYY-MM-DD` to prevent breaking charts due to above change if no `dateFormat` is provided. This change brings the code in line with the [documentation](https://mermaid.js.org/syntax/gantt.html#input-date-format)
[x] Fixed syntax for `title` and `dateFormat` in theme cypress tests. This has caused the snapshots test to fail. I have manually checked the new screenshots and the rendered chart is accurate to the diagram specified in the tests.

**Possible problems**
 - The dateRange is interpreted with the dateFormat, so if the former comes first, it'll can screw up the date parsing
   - This was mentioned on the other PR but not as a blocker from what I could tell, let me know if it is and I can look into it
 - Changing the default for `dateFormat` to `YYYY-MM-DD` may break charts that are using the ISO 8601 extended date format `YYYY-MM-DDTHH:mm:ss.sss` without specifying a `dateFormat`. This would previously have been parsed by the (now removed) backup logic using `new Date()` which would run if `dayjs` deemed the provided date to be invalid. See [comment](https://github.com/mermaid-js/mermaid/pull/5220#discussion_r1518724309).


### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [x] :computer: have added necessary unit/e2e tests.
- [x] :notebook: have added documentation. Make sure [`MERMAID_RELEASE_VERSION`](https://github.com/mermaid-js/mermaid/blob/develop/packages/mermaid/src/docs/community/contributing.md#update-documentation) is used for all new features.
- [x] :bookmark: targeted `develop` branch

paldeep-singh committed:
Add date range parser

paldeep-singh committed:
Add db functions

paldeep-singh committed:
Add db tests

paldeep-singh committed:
Add parser tests

paldeep-singh committed:
Rework renderer timescale to use new db functions

paldeep-singh committed:
Add docs update

paldeep-singh committed:
Add snapshot tests

paldeep-singh committed:
Compute start date if end-only dateRange provided

paldeep-singh committed:
Add docs for verifying snapshots

paldeep-singh committed:
Ensure tasks do not spill out of date range

paldeep-singh committed:
Remove eslint disables

paldeep-singh committed:
Fix unit tests and add tests for out of bounds dates

paldeep-singh committed:
add release version to docs

paldeep-singh committed:
Make docs clearer

paldeep-singh commented:
For the changes to the GanttDb.js and GattDb.spec.js files, I had to commit without running the pre-commit hooks as ESLint kept crashing due to Node running out of heap memory, is that a major problem?

sidharthv96 commented:
I will review the PR soon, but mentioning some ways to share credit to original authors when forking. 
1. Add their fork as [another origin](https://docs.github.com/en/get-started/getting-started-with-git/managing-remote-repositories), create a new branch, add your commits, push to your fork.
2. Add them as [Co-Authors](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors).

paldeep-singh commented:
> I will review the PR soon, but mentioning some ways to share credit to original authors when forking.
> 
>     1. Add their fork as [another origin](https://docs.github.com/en/get-started/getting-started-with-git/managing-remote-repositories), create a new branch, add your commits, push to your fork.
> 
>     2. Add them as [Co-Authors](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors).

Thanks, I did try #1, but could not pull the branch, will make sure to add a co-author to commits in the future!

sidharthv96 commented on packages/mermaid/src/diagrams/gantt/parser/gantt.spec.js:
```diff
@@ -26,6 +26,11 @@ describe('when parsing a gantt diagram it', function () {
 
     expect(parserFnConstructor(str)).not.toThrow();
   });
+  it('should handle a dateRange definition', function () {
+    const str = 'gantt\ndateRange : 2023-06-01, 2023-07-01';
+
+    expect(parserFnConstructor(str)).not.toThrow();
+  });
```
We need more test cases here. 

And the behavior below is flaky. The format mentioned in the docs is not tested here. 
So we should make it strict, then test and document it well. 

`dateRange 13:00, 14:00` fails
`dateRange  13:00, 14:00` passes (two spaces)
`dateRange: 13:00, 14:00` fails
`dateRange : 13:00, 14:00` passes

paldeep-singh replied:
I've made it more strict by removing the logic that would attempt to parse the date using `new Date` after `dayjs.isValid()` returns `false` (see [this commit](https://github.com/mermaid-js/mermaid/pull/5220/commits/81b57d6b2276f1200d5638e181165e2bd7648a62)). This change has not broken any tests, but I suspect it might break existing charts that are passing an invalid `dateFormat` while still successfully rendering due to that "backup" logic.

For valid date ranges, I've now included a test with a date-only range, a time only range, and a date+time range. Are there any other test cases I should add?

For the tests for invalid date ranges, I realised that depending on the specific error in syntax, the `dateRange` declaration could either be ignored or throw an error.

From your examples above, `dateRange: 13:00, 14:00` would be ignored since the `dateRange:` (without a space before the colon) would not be considered a match by the jison parser looking for `"dateRange"\s[^#\n;]`, and the `dateRange` would remain as the default of `''`. For `dateRange : 13:00, 14:00` (space before colon), the declaration would be picked up, but an error will be thrown as `: 13:00` would be considered an invalid date. I've added tests for both these scenarios (see [this commit](https://github.com/mermaid-js/mermaid/pull/5220/commits/0444adb25c7100cb9ae3667b440cd8caa7c8258e)), and would appreciate any feedback!

sidharthv96 commented on cypress/integration/rendering/gantt.spec.js:
```diff
@@ -523,6 +523,108 @@ describe('Gantt diagram', () => {
+      displayMode: compact
+      ---
+      gantt
+        title GANTT compact
+        dateFormat  HH:mm:ss
+        dateRange   12:30:00,13:30:00
+        axisFormat  %Hh%M
+        section DB Clean
```
Please add more test cases that handle dates as well, currently only time is used. 
Then some cases where there are tasks before and after the given range.

paldeep-singh replied:
I've started, but am new to cypress and having difficulties getting the tests to pass locally / produce screenshots that look right. Will keep going at it over the next few days and post in the Discord if I have any specific questions!

sidharthv96 commented on packages/mermaid/src/diagrams/gantt/ganttDb.js:
```diff
@@ -158,7 +208,27 @@ export const getTasks = function () {
   }
 
-  tasks = rawTasks;
+  if (dateRange === '') {
+    tasks = rawTasks;
+
+    return tasks;
+  }
```
```suggestion
  if (dateRange === '') {
    return rawTasks;
  }
```

sidharthv96 reviewed (changes requested):
Nice work @AlexMooney and @paldeep-singh. We need some more test cases and tighten up the syntax.

paldeep-singh commented:
Sorry, just been a bit busy recently, should have time in the next couple of weeks

paldeep-singh committed:
Update packages/mermaid/src/diagrams/gantt/ganttDb.js

Co-authored-by: Sidharth Vinod <sidharthv96@gmail.com>

paldeep-singh committed:
Fix dateRange parsing to not skip first character

paldeep-singh committed:
Add additional tests for valid date ranges

paldeep-singh committed:
Add tests for invalid date ranges and invalid range syntax

paldeep-singh committed:
Always throw error if dayjs determines date is invalid

paldeep-singh committed:
Improve e2e test names for time-only date ranges

paldeep-singh committed:
Add dateRange e2e tests for date-only ranges

paldeep-singh committed:
Remove compact declaration in dateRange cypress tests

paldeep-singh committed:
Add test with tasks outside of specified date range

paldeep-singh committed:
Merge branch 'develop' into feature/1530_gantt_chart_custom_range_2

paldeep-singh committed:
Fix syntax for title and dateFormat in theme tests for gantt charts

paldeep-singh committed:
Add tasks that partially exist within date range

paldeep-singh committed:
Add time-only date range tests with tasks outside range

paldeep-singh committed:
Remove tasks outside of date range in tests not testing that behaviour

paldeep-singh committed:
Default date format to YYYY-MM-DD

paldeep-singh committed:
Make chart titles in date range tests more descriptive

paldeep-singh committed:
Add e2e test for time and date dateRange with tasks extending outside of range

paldeep-singh committed:
Fix axis ticks for time-only date range test

paldeep-singh commented:
@sidharthv96 I have addressed your comments. There are 5 failing e2e tests, 1 is meant to fail and the other 4 are failures in the theme tests. I had to fix the syntax for `dateFormat` specified in the tests which has caused the chart to render differently. However, I have manually checked the new screenshots and they are accurate to the gantt diagram specified in the tests. Prior to the changes I made, the first task in the diagram wasn't rendered correctly, now it is and as such, dependent tasks have moved to the appropriate position on the chart.

paldeep-singh committed:
Merge branch 'develop' into feature/1530_gantt_chart_custom_range_2

sidharthv96 commented:
@paldeep-singh can you fix the merge conflict please?

paldeep-singh committed:
Merge branch 'develop' into feature/1530_gantt_chart_custom_range_2

paldeep-singh commented:
> @paldeep-singh can you fix the merge conflict please?

Sorry, missed that one, done now!

paldeep-singh committed:
Merge branch 'develop' into feature/1530_gantt_chart_custom_range_2

paldeep-singh commented:
@sidharthv96 Is there anything else in this PR you'd like addressed?

generrosity commented:
Sorry - was this PR ready, just pending "nothing to add"? @paldeep-singh 

Or is there pending work? @sidharthv96

paldeep-singh committed:
Merge branch 'develop' into feature/1530_gantt_chart_custom_range_2

paldeep-singh commented:
> Sorry - was this PR ready, just pending "nothing to add"? @paldeep-singh
> 
> Or is there pending work? @sidharthv96

Hi @generrosity 

I've addressed all the comments I received, just waiting on further feedback/approval. 

I just updated the branch as well, if any errors crop up in the pipeline I will have a look at them this week
